### PR TITLE
feat(*): support mixin config schema

### PR DESCRIPTION
**Pull request by vdice**
# What does this change

* Adds support for mixin config schema
  * If a mixin already declares config schema, we validate it and inject it into Porter's schema
  * If a mixin does not declare config schema, we inject a default, which allows:
    * a config-less mixin declaration, e.g.
    ```
      mixins:
        - mymixin
     ```
    * arbitrary config for the mixin (useful for when the mixin has yet to define its own schema but does already support additional configuration), e.g.
    ```
      mixins:
        - mymixin:
            arbitraryConfig: true
    ```

See also a PR for the first non-`exec` mixin to declare its configuration: https://github.com/getporter/kubernetes-mixin/pull/32

Follow-ups: If/when this all ends up looking good, we can create issues in the rest of the mixin repos in this org to define their config schema.

# What issue does it fix
Closes https://github.com/getporter/porter/issues/1341

# Notes for the reviewer

# Checklist
- [x] Unit Tests
- [ ] Documentation
- [ ] Schema (porter.yaml)

If this is your first pull request, please add your name to the bottom of our [Contributors][contributors] list. Thank you for making Porter better! 🙇‍♀️

[contributors]: https://porter.sh/src/CONTRIBUTORS.md

### CODE REVIEW

File: pkg/porter/schema.go [hunk 1]
```diff
@@ -115,32 +115,65 @@ func (p *Porter) injectMixinSchemas(manifestSchema jsonSchema) (jsonSchema, erro
...
+		}
+
+		// embed the entire mixin schema in the root after allowing for default config additions
 		manifestSchema["mixin."+mixin] = mixinSchemaMap
 
 		for _, action := range coreActions {
 			actionItemSchema, ok := actionSchemas[action]["items"].(jsonSchema)
-			if err != nil && p.Debug {
-				fmt.Fprintln(p.Err, errors.Errorf("root porter manifest schema has invalid properties.%s.items type, expected map[string]interface{} but got %T", action, actionSchemas[string(action)]["items"]))
+			if !ok && p.Debug {
```
Comment (carolynvs) [inline pkg/porter/schema.go:168]: oops

**Review by carolynvs — CHANGES_REQUESTED** [review]
Seems like whatever information that we are returning in the anyOf isn't enough for VS Code to provide proper autocomplete?

Screen Shot 2021-06-21 at 4 22 13 PM

**Comment by vdice** [comment]
Oh, shoot, I didn't think about the VS Code autocompletion for available mixin names themselves.  I was hoping to find an example where we do so, but it looks like the actions also have similar behavior.  Must be with our use of `items`... 

Screen Shot 2021-06-21 at 3 43 41 PM

**Comment by carolynvs** [comment]
Just throwing out guesses but maybe if we set title or description it will use it for display?

**Comment by vdice** [comment]
> Just throwing out guesses but maybe if we set title or description it will use it for display?

Great idea!  Have yet to find the winner from title, description, $id, but I'll keep tryin'!

**Comment by vdice** [comment]
Ok, so I still have yet to crack the nut on how to add identifiers/names to the vs code auto-complete for json schema arrays.  Would you be amenable to tackling this in a follow-up issue, as it affects/is relevant to other sections of the manifest (actions, etc.)?

I've updated the PR with changes to help the VS Code UX on auto-complete.  We add the enum list of mixin strings and place this first in the `anyOf` schema array for mixins, so that it offers a helpful prompt from the get-go.  Then, when users enter a mixin and continue to add optional configuration, the corresponding mixin config schema will be used for validation.  WDYT?


Screen Shot 2021-06-22 at 11 30 52 AM

Screen Shot 2021-06-22 at 11 31 06 AM

Screen Shot 2021-06-22 at 11 31 22 AM

Screen Shot 2021-06-22 at 12 05 02 PM

**Comment by carolynvs** [comment]
I actually still see the same funky array drop down even with the latest changes, though if I start typing a mixin, like `exTAB` I'll get `exec`.

I'm not in a hurry to get this in, the entire point is to make VS Code not be weird. So since this PR doesn't yet get us all the way there, maybe let's work with @itowlson to figure out what we need to do to make autocomplete work the way we'd like before merging anything.

**Comment by vdice** [comment]
Ah, I misunderstood the original issue.  It stated `When I declare a mixin and define configuration for it, VS Code decides that it is invalid.`  So, I focused and fixed that issue, i.e., VS Code now supports mixin configuration to be defined (it will still yell 'invalid' when it actually is invalid -- but now we actually check the mixin config schema).

I'll move this to Draft and revisit when cycles allow, since we aren't in a time crunch.

**Comment by carolynvs** [comment]
Sorry it wasn't clear! Yes, the issue is all driven by the desire for someone to be able to write a porter.yaml file with VS Code or another editor helping with autocomplete. They won't need to rely on our docs and examples as much and should be able to discover what's available and write manifests more confidently and quickly.

**Comment by itowlson** [comment]
@vdice Carolyn asked me if I could provide any advice on this.  Can you post an example of the schema for `mixins`?  I can't quite work it out from the PR itself I'm afraid.  That will help clarify whether the issue lies with the schema or with VS Code/RH YAML's handling of the schema.

**Comment by carolynvs** [comment]
I restarted with Vaughn's changes and through the power of time, trial and error, got VS code to mostly do what I wanted. Closing in favor of #2290

**Pull request closed without merging**